lib/catalog/owid/catalog/tables.py

```diff
@@ -530,3 +578,638 @@ def join(self, other: Union[pd.DataFrame, "Table"], *args, **kwargs) -> "Table":
+
+    return table
+
+
+def _add_processing_log_entry_to_each_variable(
+    table: Table, parents: List[Any], operation: variables.OPERATION
+) -> Table:
+    # Add a processing log entry to each column, including index columns.
+    for column in list(table.all_columns):
+        # New entry to add to the processing log.
+        # Note: The function add_entry_to_processing_log receives the argument variable_name, but in the processing log,
+        # the entry has the field "variable" (for simplicity).
+        log_new_entry = {"variable_name": column, "parents": parents, "operation": operation}
+
+        if log_new_entry not in table._fields[column].processing_log:
+            # If the processing log is not empty but the last entry is identical to the one we want to insert, skip, to
+            # avoid storing the same entry multiple times.
+            # This happens for example when saving tables, given that tables are stored in different formats.
+            # Otherwise, append a new entry to the processing log.
+            table._fields[column].processing_log = variables.add_entry_to_processing_log(
+                processing_log=table._fields[column].processing_log, **log_new_entry
+            )
+
+    return table
+
+
+def assign_dataset_sources_and_licenses_to_each_variable(table: Table) -> Table:
+    # Get sources and licenses from the table dataset.
+    sources = []
+    licenses = []
+    if hasattr(table.metadata, "dataset") and hasattr(table.metadata.dataset, "sources"):
+        sources = table.metadata.dataset.sources  # type: ignore
+    if hasattr(table.metadata, "dataset") and hasattr(table.metadata.dataset, "sources"):
+        licenses = table.metadata.dataset.licenses  # type: ignore
+
+    if len(sources) == len(licenses) == 0:
+        # There are no default sources/licenses to assign to each variable.
+        return table
+
+    # If a variable does not have sources/licenses defined, assign the ones from the dataset.
+    # Do this for all columns, including index columns.
+    for column in list(table.all_columns):
+        if len(table._fields[column].sources) == 0:
+            table._fields[column].sources = sources
+        if len(table._fields[column].licenses) == 0:
+            table._fields[column].licenses = licenses
+
+    return table
+
+
+@overload
+def copy_metadata(
+    from_table: Table, to_table: Table, include_missing_variables: bool = False, inplace: bool = False
+) -> Table:
+    ...
+
+
+@overload
+def copy_metadata(
+    from_table: Table, to_table: Table, include_missing_variables: bool = False, inplace: bool = True
+) -> None:
+    ...
+
+
+def copy_metadata(
+    from_table: Table, to_table: Table, include_missing_variables: bool = False, inplace: bool = False
+) -> Optional[Table]:
+    if not inplace:
+        to_table = to_table.copy()
+
+    # Copy the table metadata.
+    to_table.metadata = copy.deepcopy(from_table.metadata)
```

Comment:
FYI - I found that `copy.deepcopy` is pretty slow. Maybe you can [reuse this part of code](https://github.com/owid/etl/blob/master/lib/catalog/owid/catalog/tables.py#L452) (and make a function out of it)?

We can do it after comparing runtime of nightly run (should help us pinpoint problematic functions).

Reply:
You're right, but I'd prefer to do it in a separate PR. I've created [an isssue](https://github.com/owid/etl/issues/1269).